File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-3.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        int max_len = 0;
        int max_count = 0;
        unordered_map<char, int> mp;
        int left = 0;

        for (int right = 0; right < s.size(); right++) {
            mp[s[right]]++;
            max_count = max(max_count, mp[s[right]]);

            while ((right - left + 1) - max_count > k) {
                mp[s[left]]--;
                left++;
            }
            max_len = max(max_len, right - left + 1);
        }

        return max_len;
    }
};

File: Data Structures & Algorithms/binary-tree-diameter/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:

    best = 0

    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        
        self.maxSearch(root)

        return self.best
        
    
    def maxSearch(self, node):

        if not node:
            return 0
        
        leftBest = self.maxSearch(node.left)
        rightBest = self.maxSearch(node.right)

        self.best = max(self.best, leftBest + rightBest)

        return 1 + max(leftBest, rightBest)
